packages/core-contracts/src/contracts/arks/CrossChainArk.sol

```diff
@@ -0,0 +1,403 @@
+        config.asset.approve(address(bridgeQueue), amount);
+
+        bridgeQueue.queueTransferAssets(
+            satelliteChainId,
+            address(config.asset),
+            amount,
+            proxyAddress
+        );
+    }
+
+    /**
+     * @notice Disembarks the Ark by withdrawing assets that are available on the contract
+     * @param amount Amount of tokens to withdraw
+     * @dev This function only validates that enough assets are available on this contract
+     * The actual withdrawal from the satellite chain is processed by a keeper through
+     * FleetProxy.withdrawAndTransfer() which transfers assets back to this contract
+     */
+    function _disembark(uint256 amount, bytes calldata) internal view override {
+        // Ensure we have enough assets on the contract
+        uint256 availableAssets = config.asset.balanceOf(address(this));
+        if (availableAssets < amount) {
+            revert InsufficientAssets(amount, availableAssets);
+        }
+
+        // Note: The actual token transfer is handled by the parent Ark.disembark method
+        // No cross-chain message is required as satellite chain withdrawals are keeper-managed
+    }
+
+    /**
+     * @inheritdoc ICrossChainStateReadReceiver
+     */
+    function receiveStateRead(
+        bytes calldata resultData,
+        address requestor,
+        bytes32 requestId,
+        uint16 sourceChainId
+    ) external {
+        if (msg.sender != address(bridgeRouter)) revert Unauthorized();
+        if (sourceChainId != satelliteChainId) revert InvalidSourceChain();
+        if (requestor != address(this)) revert InvalidRequestor();
+
+        // Decode the remote asset balance
+        uint256 newRemoteBalance = abi.decode(resultData, (uint256));
+
+        lastRemoteAssetBalance = newRemoteBalance;
+
+        emit RemoteAssetBalanceUpdated(lastRemoteAssetBalance, requestId);
+
+        // Reset inflight assets as the state read now reflects the current remote balance
+        inflightAssets = 0;
+        emit InflightAssetsUpdated(0);
+    }
+
+    /**
+     * @inheritdoc ICrossChainAssetReceiver
+     */
+    function receiveMessageWithAssets(
+        address tokenAddress,
+        uint256 amount,
+        bytes calldata,
+        uint16 sourceChainId
+    ) external {
+        // Allow calls from BridgeRouter or registered bridge adapters
+        if (
+            msg.sender != address(bridgeRouter) &&
+            !bridgeRouter.isValidAdapter(msg.sender)
+        ) {
+            revert Unauthorized();
+        }
+        if (sourceChainId != satelliteChainId) revert InvalidSourceChain();
+        if (tokenAddress != address(config.asset)) revert InvalidAsset();
+
+        // Update the remote asset tracking
+        // Since we've received these assets, we can reduce the remote balance
+        if (amount <= lastRemoteAssetBalance) {
+            lastRemoteAssetBalance -= amount;
+        } else {
+            lastRemoteAssetBalance = 0;
```

Comment:
that's not necessarily true - depending on how often the updates will occur - we might have last update at 1100USDC, but the amount with yield is now 1200USDC - we send 1150USDC back.

So the actual remote assets are 50USDC - but we reset it to 0.


The receieved message could eg contain the actual balance at the time the x-chain message was sent